### ISSUE TITLE
use relative keynames in storage and export files

### ISSUE THREAD

**Issue by iandonnelly** [ElektraInitiative/libelektra#51]
- [X] simpleini
- [X] ni
- [x] tcl
- [ ] dump
- [ ] yajl with cascading keys
- [ ] xmltool (not really relevant, is superseded by xerces plugin)

I am currently having a problem with kdb import. When I try to import a file:
    cat backup.ecf | sudo kdb import system/restore
None of the keys are added from backup.ecf (which is in dump format).

If I have a file such as backup.txt in line format,
    cat backup.txt | sudo kdb import system/restore line
Works as expected.

Can anybody confirm?

**Comment by markus2330** [comment]
ni is fixed in b91ba11a9f9841db082653ed66c50e274979c088

so we still have (at least): tcl, xmltool and dump

**Comment by markus2330** [comment]
Some notes: `doc/decisions/relative.md` describes a bit more the background.

And as scattert in some todos (e.g. TOOLS) the idea is that `kdb export` does an export for each namespace, making development of plugins easier.

E.g. when you do `kdb export /hello ni` you would get following output:
```
kdb import user/hello ni << UNIQUE_EOF
relative/path/below/user/hello
...
UNIQUE_EOF
kdb import system/hello ni << UNIQUE_EOF
relative/path/below/system/hello
...
UNIQUE_EOF
```

So one can simply execute the exported file and gets everything reimported without having to care about which parts are exported with which format.
(we

As said for dump we need some compatibility so that old dump files (with absolute pathes) can still be read and even be written if not configured to use the new-style.

**Comment by markus2330** [comment]
@sanssecours do you have something ready for this PR? E.g. the fix for cascading export/import?

**Comment by sanssecours** [comment]
> @sanssecours do you have something ready for this PR? E.g. the fix for cascading export/import?

No. Otherwise I would have opened a pull request.

**Comment by markus2330** [comment]
Ok, I actually meant "which could be ready for this release" (which is unlikely to happen today or tomorrow, too many issues are still open). Maybe we should at least output a warning that cascading import/export currently is broken for non-relative plugins? Or even tag plugins if they are relative or absolute and have different import/export code. (I would prefer to avoid it, but it seems unlikely that all plugins will be fixed anytime soon, or will they?)

**Comment by sanssecours** [comment]
> Ok, I actually meant "which could be ready for this release" (which is unlikely to happen today or tomorrow, too many issues are still open).

Sorry, I do not think so. Unless the release is in a few weeks/months.

> Maybe we should at least output a warning that cascading import/export currently is broken for non-relative plugins? Or even tag plugins if they are relative or absolute and have different import/export code. (I would prefer to avoid it, but it seems unlikely that all plugins will be fixed anytime soon, or will they?)

I think tagging the sounds like a good idea. Although, ultimately I think we should just fix this issue. You are right though: I assume it will take quite some time until I am done with this.

**Comment by markus2330** [comment]
Ok, thanks for the update. You might over-estimate it: Once you have done one of the plugins, the others are really similar.

**Comment by markus2330** [comment]
This is also a serious issue for the web service, where the real path does not match what the user believes the absolute path is.

**Comment by markus2330** [comment]
simpleini should be fixed in 665477c6f93430aa803c5831f2bd9673440cb667, see head of post for summary.

**Comment by markus2330** [comment]
I close this issue as dump and xmltool are soon only relevant as legacy formats. (See #2330 #1327)
